preserve signatures with multiple calls + buffer mutations

avikchaudhuri opened the pull request:
As called out in https://github.com/pytorch/pytorch/pull/137999, preserving signatures of multiple calls when buffer mutations are present was NYI. The main problem was that intermediate values of buffers were not tracked, so couldn't be propagated statefully between multiple calls (i.e., they would need to be explicitly passed around, defeating the unlifting needed for preserving signatures). 

This PR fixes this situation, by introducing module attributes that carry the necessary intermediate values of buffer mutations. In general, a buffer mutation can have several intermediate values it depends on recursively, even other buffers. So rather than tying an intermediate value with a particular buffer, we tie it with the submodules that create and read it. We install an attribute on all modules that create or read a particular intermediate value, sharing the same initial storage (i.e., initialized with the same empty tensor). For the module that creates this intermediate value, we copy the value into the corresponding attribute; and for the modules that read it, we read the corresponding attribute instead.

Another complication that needed to be addressed was that a `run_decompositions` following an `export_for_training` was not preserving module call graphs, which is needed for unflattening and, in particular, used when remapping inputs. Fortunately some existing metadata already tracks provenance of nodes, which we could use to update a module call graph after functionalization / decomposition.

Differential Revision: D64806175

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -6575,25 +6575,25 @@ def forward(self, x):
+            test(export(M(), inp, preserve_module_call_signature=("n",)))
+            if is_training_ir_test(self._testMethodName):
+                # export_for_training + run_decompositions
+                ep = export(M(), inp, preserve_module_call_signature=("n",))
```
Can you also test unflatten on training_ir directly?

tugsbayasgalan commented on torch/export/exported_program.py:
```diff
@@ -643,6 +643,30 @@ def _common_getitem_elimination_pass(
+    # from old node names to new node names
+    provenance: Dict[str, str] = {}
+    for node in gm.graph.nodes:
+        if history := node.meta.get("from_node", []):
```
What does "from_node" mean actually? Shouldn't we also need to rewrite the node.meta after run_decomp to reflect the change in "from_node"?

avikchaudhuri replied:
"from_node" keeps a history of how a node was generated from tracing other nodes.

avikchaudhuri replied:
So after the first decomp, it will contain the original node from export; after the second decomp, it will contain the original node followed by the node in the first decomp; etc.

tugsbayasgalan commented on torch/export/unflatten.py:
```diff
@@ -584,7 +579,74 @@ def unflatten(
+    assert output_node.op == "output" and len(output_node.args) == 1
+    return_args = output_node.args[0]
+
+    mutation_node_to_buffer = graph_signature.buffers_to_mutate
```
Hmm this won't work for training IR cause because we don't have this information...

tugsbayasgalan replied:
And fixing this is probably complicated, maybe we should document it in the unflatten function, could you file a task to track this?

avikchaudhuri replied:
Note that this is only needed when buffer updates have been functionalized away. If the code contains direct buffer updates, then none of this is required. So it should be fine.

avikchaudhuri replied:
Hmm, I was wrong. It turns out that the output node, rather than the input node, of mutation is what is threaded through the rest of the program, so every buffer mutation won't have the placeholder node corresponding to the buffer.

I think temporarily decomposing the training IR is overkill. How can I detect mutating ops?

avikchaudhuri committed:
preserve signatures with multiple calls + buffer mutations (#138669)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/138669

Test Plan: modified test

Differential Revision: D64806175

tugsbayasgalan commented on torch/export/unflatten.py:
```diff
@@ -1198,6 +1203,106 @@ def _reorder_submodules(
+            for fqn in fqns:
+                ivals[fqn].append(name)
+
+        for entry in entries:
```
Why are we creating intermediate value for ALL seen modules so far? Should't we only do it if the check on line 1285 passes?

avikchaudhuri replied:
Not for all seen modules, just the ones that satisfy the check on line 1285. For those intermediate values, we have the fqns for modules that read or update them, so we specifically initialize the attributes for this subset of fqns.

avikchaudhuri commented:
@pytorchbot merge -f "Landed internally"
